iphone/TitaniumKit/TitaniumKit/Sources/API/ObjcProxy.m

```diff
@@ -0,0 +1,235 @@
+        // if the callback matches, remove the listener from our mapping and mark unmanaged
+        [actualCallback.context.virtualMachine removeManagedReference:storedCallback withOwner:self];
+        [listenersForType removeObjectAtIndex:i];
+        [m_listeners setObject:listenersForType forKey:name];
+        ourCallbackCount = count - 1;
+        removed = true;
+        break;
+      }
+    }
+  }
+  @finally {
+    pthread_rwlock_unlock(&m_listenerLock);
+    if (removed) {
+      [self _listenerRemoved:name count:ourCallbackCount];
+    }
+  }
+}
+
+- (BOOL)_hasListeners:(NSString *)type
+{
+  pthread_rwlock_rdlock(&m_listenerLock);
+  @try {
+    if (m_listeners == nil) {
+      return NO;
+    }
+    NSMutableArray *listenersForType = (NSMutableArray *)[m_listeners objectForKey:type];
+    if (listenersForType == nil) {
+      return NO;
+    }
+    return [listenersForType count] > 0;
+  }
+  @finally {
+    pthread_rwlock_unlock(&m_listenerLock);
+  }
+}
+
+- (void)_listenerAdded:(NSString *)type count:(int)count
+{
+  // for subclasses
+}
+
+- (void)_listenerRemoved:(NSString *)type count:(int)count
+{
+  // for subclasses
+}
+
+- (void)fireEvent:(NSString *)name withDict:(NSDictionary *)dict
+{
+  pthread_rwlock_rdlock(&m_listenerLock);
+  @try {
+    if (m_listeners == nil) {
+      return;
+    }
+    NSArray *listenersForType = [m_listeners objectForKey:name];
+    if (listenersForType == nil) {
+      return;
+    }
+    // FIXME: looks like we need to handle bubble logic/etc. See other fireEvent impl
+    for (JSManagedValue *storedCallback in listenersForType) {
+      JSValue *function = [storedCallback value];
+      [function callWithArguments:@[ dict ]];
+    }
+  }
+  @finally {
+    pthread_rwlock_unlock(&m_listenerLock);
+  }
+}
+
+- (void)_fireEventToListener:(NSString *)type withObject:(id)obj listener:(JSValue *)listener
+{
+  NSMutableDictionary *eventObject = nil;
+  if ([obj isKindOfClass:[NSDictionary class]]) {
+    eventObject = [NSMutableDictionary dictionaryWithDictionary:obj];
+  } else {
+    eventObject = [NSMutableDictionary dictionary];
+  }
+
+  // common event properties for all events we fire.. IF they're undefined.
+  if (eventObject[@"type"] == nil) {
+    eventObject[@"type"] = type;
+  }
+  if (eventObject[@"source"] == nil) {
+    eventObject[@"source"] = self;
+  }
+
+  if (listener != nil) {
+    [listener callWithArguments:@[ eventObject ]];
+  }
+}
+
+//For subclasses to override
+- (NSString *)apiName
+{
+  DebugLog(@"[ERROR] Subclasses must override the apiName API endpoint.");
+  return @"Ti.Proxy";
+}
+
+GETTER_IMPL(NSString *, apiName, ApiName);
+READWRITE_IMPL(BOOL, bubbleParent, BubbleParent);
+
+- (void)dealloc
+{
+  [self _destroy];
+  pthread_rwlock_destroy(&m_listenerLock);
+  [super dealloc];
+}
+
+- (void)_destroy
+{
+  // remove all listeners JS side proxy
+  pthread_rwlock_wrlock(&m_listenerLock);
+  // FIXME: Need to unhoook managed values!
```

Comment:
The docs state that if the `JSManagedValue` is not reachable anymore from the ObjC object graph it releases its underlying `JSValue`. So there should be no need to explicitly remove the object reference.

Reply:
ok